### PR TITLE
[Feature] x_dot_x builtin kernel support

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Adding support for u\_dot\_v, u\_dot\_e, v\_dot\_e, v\_dot\_u, e\_dot\_u and e\_dot\_v  as builtin kernel.
Implementation will base on current binary_reduce structure.
#659

Tasks
- [x] Forward of [x]\_add\_[x], [x]\_sub\_[x], [x]\_mul\_[x], [x]\_div\_[x] still work
- [x] Backward of [x]\_add\_[x], [x]\_sub\_[x], [x]\_mul\_[x], [x]\_div\_[x] still work
- [x] Forward of [x]\_dot\_[x]
- [x] Backward of [x]\_dot\_[x]

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by VoVAllen** [comment]
MXNet CI test may have some problem due to adapting to new version. I'll try to fix this tomorrow.

File: python/dgl/function/message.py [hunk 1]
```diff
@@ -147,12 +147,16 @@ def copy_e(e, out):
...
+# element-wise message functions:
 # u_add_v, u_sub_v, u_mul_v, u_div_v
 # u_add_e, u_sub_e, u_mul_e, u_div_e
 # v_add_u, v_sub_u, v_mul_u, v_div_u
 # v_add_e, v_sub_e, v_mul_e, v_div_e
 # e_add_u, e_sub_u, e_mul_u, e_div_u
 # e_add_v, e_sub_v, e_mul_v, e_div_v
+#
+# dot message functions:
+# u_dot_v, u_dot_e, v_dot_e
```
Comment (yzh119) [inline python/dgl/function/message.py]: Please also provide `v_dot_u`, `e_dot_u` and `e_dot_v`

File: src/kernel/binary_reduce_common.h [hunk 2]
```diff
@@ -192,6 +193,24 @@ struct BinaryUseLhs {
   }
 };
 
+template <typename DType>
+struct BinaryDot {
+  static DGLDEVICE DGLINLINE DType Call(DType *lhs, DType *rhs, int64_t len) {
+    DType out = 0;
+    // simple vector dot vector
```
Comment (yzh119) [inline src/kernel/binary_reduce_common.h]: We can unroll the loop here.

File: src/kernel/cpu/binary_reduce_impl.h [hunk 3]
```diff
@@ -92,17 +92,17 @@ struct BinaryReduceBcast {
...
     int64_t tmp[NDim];  // store unraveled idx.
     for (int64_t tx = 0; tx < gdata->out_len; ++tx) {
       Unravel(tx, gdata->ndim, gdata->out_shape, gdata->out_stride, tmp);
-      DType lhs = Functors::Read(lhsoff +
-          Ravel(tmp, gdata->ndim, gdata->lhs_shape, gdata->lhs_stride));
-      DType rhs = Functors::Read(rhsoff +
-          Ravel(tmp, gdata->ndim, gdata->rhs_shape, gdata->rhs_stride));
-      DType out = Functors::Op(lhs, rhs);
+      DType out = Functors::Op(
+          lhsoff + Ravel(tmp, gdata->ndim, gdata->lhs_shape, gdata->lhs_stride) * len,
```
Comment (yzh119) [inline src/kernel/cpu/binary_reduce_impl.h]: Loop unroll for ravel?

**Comment by yzh119** [comment]
@jermainewang , I've verified the correctness with STT. The GPU memory footprint is about the same but builtin function is 2x slower than my custom kernels(node parallel strategy).

File: src/kernel/cpu/binary_reduce_impl.h (hunk 3 above)
Comment (classicsong) [reply src/kernel/cpu/binary_reduce_impl.h]: > Loop unroll for ravel?

We can opt this, after another code is merged, by fusing Unravel and Ravel

File: python/dgl/function/message.py (hunk 1 above)
Comment (classicsong) [reply python/dgl/function/message.py]: Done

File: src/kernel/cpu/binary_reduce_impl.h (hunk 3 above)
Comment (classicsong) [reply src/kernel/cpu/binary_reduce_impl.h]: I have applied UnravelRavel opt for both forward and backward

**Pull request merged**